go/stellar/acctbundle/acctbundle.go

```diff
@@ -0,0 +1,673 @@
+	if err != nil {
+		return stellar1.EncryptedBundle{}, "", err
+	}
+	secbox := secretbox.Seal(nil, clearpack[:], &nonce, (*[libkb.NaclSecretBoxKeySize]byte)(&symmetricKey))
+
+	// Annotate
+	res := stellar1.EncryptedBundle{
+		V:   2,
+		E:   secbox,
+		N:   nonce,
+		Gen: pukGen,
+	}
+
+	// Msgpack (outer) + b64
+	cipherpack, err := libkb.MsgpackEncode(res)
+	if err != nil {
+		return stellar1.EncryptedBundle{}, "", err
+	}
+	resB64 := base64.StdEncoding.EncodeToString(cipherpack)
+	return res, resB64, nil
+}
+
+// AcctBoxedEncoded is the result of boxing and encoding the per-account secrets.
+type AcctBoxedEncoded struct {
+	Enc           stellar1.EncryptedAccountBundle
+	EncHash       stellar1.Hash
+	EncB64        string // base64 msgpacked Enc
+	FormatVersion stellar1.AccountBundleVersion
+}
+
+func accountBoxAndEncode(accountID stellar1.AccountID, accountBundle stellar1.AccountBundle, pukGen keybase1.PerUserKeyGeneration, puk libkb.PerUserKeySeed) (*AcctBoxedEncoded, error) {
+	versionedSecret := stellar1.NewAccountBundleSecretVersionedWithV1(stellar1.AccountBundleSecretV1{
+		AccountID: accountID,
+		Signers:   accountBundle.Signers,
+	})
+
+	encBundle, b64, err := accountEncrypt(versionedSecret, pukGen, puk)
+	if err != nil {
+		return nil, err
+	}
+
+	encPack, err := libkb.MsgpackEncode(encBundle)
+	if err != nil {
+		return nil, err
+	}
+	encHash := sha256.Sum256(encPack)
+
+	res := AcctBoxedEncoded{Enc: encBundle, EncHash: encHash[:], EncB64: b64, FormatVersion: 1}
+
+	return &res, nil
+}
+
+// ErrNoChangeNecessary means that any proposed change to a bundle isn't
+// actually necessary.
+var ErrNoChangeNecessary = errors.New("no account mode change is necessary")
+
+// MakeMobileOnly transforms a stellar1.AccountBundle into a mobile-only
+// bundle.  This advances the revision.  If it's already mobile-only,
+// this function will return ErrNoChangeNecessary.
+func MakeMobileOnly(a *stellar1.BundleRestricted, accountID stellar1.AccountID) error {
+	ws, err := AccountWithSecret(a, accountID)
+	if err != nil {
+		return err
+	}
+	if ws.Mode == stellar1.AccountMode_MOBILE {
+		return ErrNoChangeNecessary
+	}
+	a.Revision++
+	a.Prev = a.OwnHash
+	a.OwnHash = nil
+
+	for i, vis := range a.Accounts {
+		if vis.AccountID == accountID {
+			if vis.Mode == stellar1.AccountMode_MOBILE {
+				return ErrNoChangeNecessary
+			}
+			vis.Mode = stellar1.AccountMode_MOBILE
+			vis.AcctBundleRevision++
+			a.Accounts[i] = vis
+		}
+	}
+
+	return nil
+}
+
+// MakeAllDevices transforms a stellar1.AccountBundle into an all-device
+// bundle.  This advances the revision.  If it's already all-device,
+// this function will return ErrNoChangeNecessary.
+func MakeAllDevices(a *stellar1.BundleRestricted, accountID stellar1.AccountID) error {
+	ws, err := AccountWithSecret(a, accountID)
+	if err != nil {
+		return err
+	}
+	if ws.Mode == stellar1.AccountMode_USER {
+		return ErrNoChangeNecessary
+	}
+	a.Revision++
+	a.Prev = a.OwnHash
+	a.OwnHash = nil
```

Comment:
BundleRestricted is very similar to Bundle. So there's a bunch of duplicated code like these assignments in MakeAllDevices and MakeMobileOnly. What about just using Bundle and then using bundle.Advance for these?

Reply:
But it's different (BundleRestricted).  

I felt it was simpler to leave Bundle alone than try to make it work with existing bundles and new bundles.